tests/ui/eta.stderr

```diff
@@ -24,12 +24,6 @@ error: redundant closure
 LL |     all(&[1, 2, 3], &&2, |x, y| below(x, y)); //is adjusted
    |                          ^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `below`
 
-error: redundant closure
-  --> $DIR/eta.rs:40:27
-   |
-LL |     let e = Some(1u8).map(|a| divergent(a));
-   |                           ^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `divergent`
-
```

Comment:
This seems like a false negative because `map` allows a function with *any* return type but the bug in the example expects a function that returns nothing (or unit). It seems like it should be possible to detect the difference. Maybe there is an expr type adjustment to coerce `!` to `()` on the closure or the function call within the closure. Or maybe you can detect this scenario by comparing `call_ty` and `closure_ty` (one of them has `-> ()` and the other has `-> !`?).

Reply:
Ah, I definitely misunderstood the bug in the example. I will take a further look

Reply:
Seems like checking for coerced closure is not enough. In the case above, it seems like the closure is coerced to `()` before being passed on to `map` (not sure why this is the case, though). AFAIU, in the example above, `map` accepts `divergent` nonetheless, probably because of generics. For non-generics, this is not the case and type check wouldn't resolve (hence the problem with issue #8416 and similar, see [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ee3ed9301b745e22ceba0406588b0bc6)).

Ignoring the coerced closure is not hard, but it does produce false negatives such as the case above. Although I'd say that this is still an improvement over all the false positives.

A full solution would be to check if type-check is resolvable for the call/method-call taking  `Fn(..) -> T` without the coerced closure, but I'm not sure that this is trivial to do.